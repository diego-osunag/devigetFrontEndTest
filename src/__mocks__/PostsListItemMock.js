const PostsListItemMock = {
  'id': 'cid9999',
  'visited': false,
  'author': 'chrono_trigger416',
  'created_utc': 1623330764,
  'thumbnail': 'https://a.thumbs.redditmedia.com/08MUc3eyJ4zs2f29UIAG53tGKMU6KwOFTb1An2aGUJ4.jpg',
  'title': 'Lorem ipsum dolor sit amet andenum canus lupus familiaris est ergo.',
  'num_comments': 99,
};

export default PostsListItemMock;
