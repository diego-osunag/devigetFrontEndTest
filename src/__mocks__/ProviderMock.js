import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import stateMock from './stateMock';
import reducer from '../reducers';

const store = createStore(reducer, stateMock);

const ProviderMock = (props) => (
  <Provider store={store}>
    {props.children}
  </Provider>
);

export default ProviderMock;
