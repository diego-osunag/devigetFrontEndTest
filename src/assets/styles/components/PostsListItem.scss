.postsListItem {
  color: #f5f5f5;
  border-bottom: 1px solid #666;
  display: flex;
  flex-direction: column;
  margin: 0 6px;
  cursor: pointer;
  z-index: 1;
  overflow: hidden;
  position: relative;
}

.postListItem:hover {
  background-color: red;
}

.postsListItem__row {
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
  padding: 10px 0 10px 0;
}

.postsListItem__author {
  display: flex;
}

.postsListItem__author--indicator {
  width: 15px;
  height: auto;
  margin-right: 12px;
  align-self: center;
  transition: opacity 300ms;
}

.postsListItem__author--indicator.visited {
  opacity: 0;
}

.postsListItem__author--text {
  color: white;
  font-size: 1.2em;
  font-weight: normal;
  margin: 0 25px 0 0;
  transition: color 300ms;
  overflow: hidden;
  position: relative;
}

.postsListItem__author--text.visited {
  color: #999;
}

.postsListItem__time {
  justify-self: flex-start;
  text-align: right;
  color: #999;
  margin: 0 0 1px 0;
  font-size: 1.1em;
  padding-right: 22px;
  width: 155px;
}

.postsListItem__thumb {
  max-width: 100px;
  height: auto;
  align-self: center;
}

.postsListItem__description {
  text-align: left;
  margin: 0;
  padding: 5px 11px;
  flex-grow: 1;
  align-self: center;
  width: 100%;
  transition: color 300ms;
}

.postsListItem__description.visited {
  color: #999;
}

.postsListItem__icon {
  width: 15px;
  height: auto;
  align-self: center;
}

.postsListItem__dismiss {
  display: flex;
  align-items: center;
  justify-content: center;
  height: auto;
  align-self: center;
  cursor: pointer;
  padding: 5px 10px;
  border-radius: 5px;
  transition: background-color 400ms;
  z-index: 2;
}

.postsListItem__dismiss:hover {
  background-color: #333;
}

.postsListItem__dismiss:active {
  background-color: #eee;
}

.postsListItem__dismiss--icon {
  width: 15px;
  height: auto;
  margin-right: 7px;
}

.postsListItem__dismiss--text {
  margin: 0;
}

.postsListItem__comments {
  color: #f3bb83;
  padding-right: 22px;
  margin: 0;
  padding-bottom: 5px;
}