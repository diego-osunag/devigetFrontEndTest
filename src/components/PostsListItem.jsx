import React from 'react';
import { connect } from 'react-redux';
import { setSelectedPost, removeSelectedPost, dismissPost, setPostAsVisited, dismissAllPosts } from '../actions';
import '../assets/styles/components/PostsListItem.scss';
import dotIcon from '../assets/static/circle.png';
import arrowIcon from '../assets/static/right-arrow-icon.png';
import dismissIcon from '../assets/static/dismiss-icon.png';

const PostsListItem = (post) => {
  const handlePostClick = (itemId) => {
    if (!post.visited) post.setPostAsVisited(itemId);
    post.setSelectedPost(itemId);
  };

  const handleDismissPost = (itemId) => {
    post.removeSelectedPost();
    post.dismissPost(itemId);
  };

  const calcTimeStamp = () => {
    const today = new Date();
    const currentTime = Math.floor(today.getTime() / 1000);

    const timeDifference = currentTime - post.created_utc;
    let differenceConverted;
    let timeUnit = '';

    if (timeDifference > 3600) {
      differenceConverted = Math.floor(timeDifference / 3600);
      timeUnit = differenceConverted > 1 ? 'hours' : 'hour';

    } else if (timeDifference > 60) {
      differenceConverted = Math.floor(timeDifference / 60);
      timeUnit = differenceConverted > 1 ? 'minutes' : 'minute';

    } else {
      differenceConverted = timeDifference;
      timeUnit = differenceConverted > 1 ? 'seconds' : 'second';
    }

    return `${differenceConverted} ${timeUnit} ago`;
  };

  const validateThumbUrl = () => {
    switch (post.thumbnail) {
      case 'self':
        return '';
      case 'default':
        return '';
      default:
        return post.thumbnail;
    }
  };

  return (
    <div className='postsListItem'>
      <div className='postsListItem__row' onClick={() => handlePostClick(post.id)}>
        <div className='postsListItem__author'>
          <img
            className={post.visited ? 'postsListItem__author--indicator visited' : 'postsListItem__author--indicator'}
            src={dotIcon}
            alt='Blue dot'
          />
          <h2
            className={post.visited ? 'postsListItem__author--text visited' : 'postsListItem__author--text' }
          >
            {post.author}
          </h2>
        </div>
        <p className='postsListItem__time'>{ calcTimeStamp() }</p>
      </div>
      <div className='postsListItem__row' onClick={() => handlePostClick(post.id)}>
        <img
          className={post.visited ? 'postsListItem__thumb visited' : 'postsListItem__thumb'}
          src={validateThumbUrl()}
          alt=''
        />
        <p className={ post.visited ? 'postsListItem__description visited' : 'postsListItem__description' }>
          { post.title }
        </p>
        <img className='postsListItem__icon' src={ arrowIcon } alt='Right Arrow Icon' />
      </div>
      <div className='postsListItem__row'>
        <div className='postsListItem__dismiss'  onClick={() => handleDismissPost(post.id)}>
          <img className='postsListItem__dismiss--icon' src={ dismissIcon } alt='Dismiss Icon' />
          <p className='postsListItem__dismiss--text'>Dismiss Post</p>
        </div>
        <p className='postsListItem__comments' onClick={() => handlePostClick(post.id)}>{`${post.num_comments} comments`}</p>
      </div>
    </div>
  );
};

const mapDispatchToProps = {
  setSelectedPost, removeSelectedPost, dismissPost, setPostAsVisited, dismissAllPosts,
};

export default connect(null, mapDispatchToProps)(PostsListItem);