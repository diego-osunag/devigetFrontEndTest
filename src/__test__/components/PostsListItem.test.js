import React from 'react';
import { mount } from 'enzyme';
import { create } from 'react-test-renderer';
import ProviderMock from '../../__mocks__/ProviderMock';
import PostsListItemMock from '../../__mocks__/PostsListItemMock';
import PostsListItem from '../../components/PostsListItem';

describe('<PostsListItem />', () => {
  const post = mount(
    <ProviderMock>
      <PostsListItem
        id={PostsListItemMock.id}
        author={PostsListItemMock.author}
        created_utc={PostsListItemMock.created_utc}
        thumbnail={PostsListItemMock.thumbnail}
        title={PostsListItemMock.title}
        num_comments={PostsListItemMock.num_comments}
      />
    </ProviderMock>,
  );

  test('PostsListItem component rendering', () => {
    expect(post.length).toEqual(1);
  });

  test('Read/Unread indicator rendering', () => {
    expect(post.find('img.postsListItem__author--indicator').length).toEqual(1);
  });

  test('Author\'s name rendering', () => {
    expect(post.find('h2.postsListItem__author--text').text()).toEqual('chrono_trigger416');
  });

  test('Timestamp label rendering (X hours ago)', () => {
    expect(post.find('.postsListItem__time').length).toEqual(1);
  });

  test('Timestamp label rendering (X hour ago)', () => {
    const oneHourAgo = () => {
      const today = new Date();
      const currentTime = Math.floor(today.getTime() / 1000);
      return (currentTime - 3600);
    };
    const post1 = mount(
      <ProviderMock>
        <PostsListItem
          created_utc={oneHourAgo}
        />
      </ProviderMock>,
    );
    expect(post1.find('.postsListItem__time').length).toEqual(1);
  });

  test('Timestamp label rendering (X minutes ago)', () => {
    const thirtyMinutesAgo = () => {
      const today = new Date();
      const currentTime = Math.floor(today.getTime() / 1000);
      return (currentTime - 1800);
    };
    const post2 = mount(
      <ProviderMock>
        <PostsListItem
          created_utc={thirtyMinutesAgo}
        />
      </ProviderMock>,
    );
    expect(post2.find('.postsListItem__time').length).toEqual(1);
  });

  test('Timestamp label rendering (X minute ago)', () => {
    const oneMinuteAgo = () => {
      const today = new Date();
      const currentTime = Math.floor(today.getTime() / 1000);
      return (currentTime - 60);
    };
    const post3 = mount(
      <ProviderMock>
        <PostsListItem
          created_utc={oneMinuteAgo}
        />
      </ProviderMock>,
    );
    expect(post3.find('.postsListItem__time').length).toEqual(1);
  });

  test('Timestamp label rendering (X seconds ago)', () => {
    const tenSecondsAgo = () => {
      const today = new Date();
      const currentTime = Math.floor(today.getTime() / 1000);
      return (currentTime - 10);
    };
    const post4 = mount(
      <ProviderMock>
        <PostsListItem
          created_utc={tenSecondsAgo}
        />
      </ProviderMock>,
    );
    expect(post4.find('.postsListItem__time').length).toEqual(1);
  });

  test('Timestamp label rendering (X second ago)', () => {
    const oneSecondAgo = () => {
      const today = new Date();
      const currentTime = Math.floor(today.getTime() / 1000);
      return (currentTime - 1);
    };
    const post5 = mount(
      <ProviderMock>
        <PostsListItem
          created_utc={oneSecondAgo}
        />
      </ProviderMock>,
    );
    expect(post5.find('.postsListItem__time').length).toEqual(1);
  });

  test('Thumbnail image rendering', () => {
    expect(post.find('.postsListItem__thumb').length).toEqual(1);
  });

  test('Post title rendering', () => {
    expect(post.find('.postsListItem__description').text()).toEqual('Lorem ipsum dolor sit amet andenum canus lupus familiaris est ergo.');
  });

  test('Right arrow image rendering', () => {
    expect(post.find('.postsListItem__icon').length).toEqual(1);
  });

  test('Dismiss button rendering', () => {
    expect(post.find('.postsListItem__dismiss').length).toEqual(1);
  });

  test('Post number of comments rendering', () => {
    expect(post.find('.postsListItem__comments').text()).toEqual('99 comments');
  });
});

describe('<PostsListItem /> Snapshot', () => {
  test('PostsListItem UI rendering correctly', () => {
    const post = create(
      <ProviderMock>
        <PostsListItem />
      </ProviderMock>,
    );
    expect(post.toJSON()).toMatchSnapshot();
  });
});
