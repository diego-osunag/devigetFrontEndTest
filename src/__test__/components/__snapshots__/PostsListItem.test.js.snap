// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostsListItem /> Snapshot PostsListItem UI rendering correctly 1`] = `
<div
  className="postsListItem"
>
  <div
    className="postsListItem__row"
    onClick={[Function]}
  >
    <div
      className="postsListItem__author"
    >
      <img
        alt="Blue dot"
        className="postsListItem__author--indicator"
        src={Object {}}
      />
      <h2
        className="postsListItem__author--text"
      />
    </div>
    <p
      className="postsListItem__time"
    >
      NaN second ago
    </p>
  </div>
  <div
    className="postsListItem__row"
    onClick={[Function]}
  >
    <img
      alt=""
      className="postsListItem__thumb"
    />
    <p
      className="postsListItem__description"
    />
    <img
      alt="Right Arrow Icon"
      className="postsListItem__icon"
      src={Object {}}
    />
  </div>
  <div
    className="postsListItem__row"
  >
    <div
      className="postsListItem__dismiss"
      onClick={[Function]}
    >
      <img
        alt="Dismiss Icon"
        className="postsListItem__dismiss--icon"
        src={Object {}}
      />
      <p
        className="postsListItem__dismiss--text"
      >
        Dismiss Post
      </p>
    </div>
    <p
      className="postsListItem__comments"
      onClick={[Function]}
    >
      undefined comments
    </p>
  </div>
</div>
`;
